python/pylibcudf/pylibcudf/tests/test_hashing.py

```diff
@@ -0,0 +1,186 @@
+        hasher = mmh3.mmh3_x64_128(seed=plc.hashing.LIBCUDF_DEFAULT_HASH_SEED)
+        hasher.update(x)
+        # libcudf returns a tuple of two 64-bit integers
+        return hasher.utupledigest()
+    elif method == "xxhash_64":
+        return xxhash.xxh64(
+            x, seed=plc.hashing.LIBCUDF_DEFAULT_HASH_SEED
+        ).intdigest()
+    else:
+        return getattr(hashlib, method)(x).hexdigest()
+
+
+@pytest.mark.parametrize(
+    "method", ["sha1", "sha224", "sha256", "sha384", "sha512"]
+)
+def test_hash_column_sha(pa_scalar_input_column, plc_scalar_input_tbl, method):
+    plc_hasher = getattr(plc.hashing, method)
+
+    def py_hasher(val):
+        return getattr(hashlib, method)(scalar_to_binary(val)).hexdigest()
+
+    expect = pa.array(
+        [py_hasher(val) for val in pa_scalar_input_column.to_pylist()],
+        type=pa.string(),
+    )
+    got = plc_hasher(plc_scalar_input_tbl)
+    assert_column_eq(got, expect)
+
+
+def test_hash_column_md5(pa_scalar_input_column, plc_scalar_input_tbl):
+    def py_hasher(val):
+        return hashlib.md5(scalar_to_binary(val)).hexdigest()
+
+    expect = pa.array(
+        [py_hasher(val) for val in pa_scalar_input_column.to_pylist()],
+        type=pa.string(),
+    )
+    got = plc.hashing.md5(plc_scalar_input_tbl)
+    assert_column_eq(got, expect)
+
+
+def test_hash_column_xxhash64(pa_scalar_input_column, plc_scalar_input_tbl):
+    def py_hasher(val):
+        return xxhash.xxh64(
+            scalar_to_binary(val), seed=plc.hashing.LIBCUDF_DEFAULT_HASH_SEED
+        ).intdigest()
+
+    expect = pa.array(
+        [py_hasher(val) for val in pa_scalar_input_column.to_pylist()],
+        type=pa.uint64(),
+    )
+    got = plc.hashing.xxhash_64(plc_scalar_input_tbl, 0)
+
+    assert_column_eq(got, expect)
+
+
+@pytest.mark.parametrize(
+    "method", ["sha1", "sha224", "sha256", "sha384", "sha512"]
+)
+@pytest.mark.parametrize("dtype", ["list", "struct"])
+def test_sha_list_struct_err(list_struct_table, dtype, method):
+    err_types = list_struct_table.select([dtype])
+    plc_tbl = plc.interop.from_arrow(err_types)
+    plc_hasher = getattr(plc.hashing, method)
+
+    with pytest.raises(TypeError):
+        plc_hasher(plc_tbl)
+
+
+def test_md5_struct_err(list_struct_table):
+    err_types = list_struct_table.select(["struct"])
+    plc_tbl = plc.interop.from_arrow(err_types)
+
+    with pytest.raises(TypeError):
+        plc.hashing.md5(plc_tbl)
+
+
+def test_murmurhash3_x86_32(pa_scalar_input_column, plc_scalar_input_tbl):
+    def py_hasher(val):
+        return libcudf_mmh3_x86_32(scalar_to_binary(val))
+
+    got = plc.hashing.murmurhash3_x86_32(plc_scalar_input_tbl, 0)
+    expect = pa.array(
+        [py_hasher(val) for val in pa_scalar_input_column.to_pylist()],
+        type=pa.uint32(),
+    )
+    got = plc.hashing.murmurhash3_x86_32(plc_scalar_input_tbl, 0)
+    assert_column_eq(got, expect)
+
+
+# def test_murmurhash_x86_32_list_struct TODO
```

Comment:
Here's a missing test. I will take another whack and coming up with a python equivalent, if desired.

Reply:
For lists, it should be something like:

```python
hash_value = 0
seed = plc.hashing.LIBCUDF_DEFAULT_HASH_SEED
for element in list_data:
    hash_value = hash_combine_32(hash_value, mmh3.hash(element, seed))
```

https://github.com/rapidsai/cudf/blob/4fe338c0efe0fee2ee69c8207f9f4cbe9aa4d4a2/cpp/include/cudf/table/experimental/row_operators.cuh#L1930-L1938


For structs, this will be more complicated. We "decompose" structs before hashing them. https://github.com/rapidsai/cudf/blob/4fe338c0efe0fee2ee69c8207f9f4cbe9aa4d4a2/cpp/src/table/row_operators.cu#L96

I am pretty sure that this can be implemented with a recursive algorithm in Python, but we could not implement the CUDA kernel that way for performance reasons. Therefore, the [hash row operator for structs](https://github.com/rapidsai/cudf/blob/4fe338c0efe0fee2ee69c8207f9f4cbe9aa4d4a2/cpp/include/cudf/table/experimental/row_operators.cuh#L1926-L1929) in libcudf won't help you understand what to write here.

However, I don't think that we _need_ to implement recursive struct tests here. I would stick to testing one-level structs, which would just mean that you need a simple loop like we used for the lists.

```python
hash_value = 0
seed = plc.hashing.LIBCUDF_DEFAULT_HASH_SEED
for element in struct_data:
    hash_value = hash_combine_32(hash_value, mmh3.hash(element, seed))
```